layout: Rewrite anonymous table code, simplify and fix table intrinsic width calculation, and improve safety of flexbox code.

pcwalton opened the pull request:
Closes #13782.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13870)

<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/selector_matching.rs, components/style/animation.rs, components/style/servo_selector_impl.rs, components/style/properties/properties.mako.rs, components/style/matching.rs
 * @kichjang: components/script_layout_interface/wrapper_traits.rs
 * @fitzgen: components/script_layout_interface/wrapper_traits.rs
 * @emilio: components/style/selector_matching.rs, components/style/animation.rs, components/style/servo_selector_impl.rs, components/layout/multicol.rs, components/layout/list_item.rs, components/style/properties/properties.mako.rs, components/layout/table_caption.rs, components/layout/flex.rs, components/layout/table_cell.rs, components/layout/construct.rs, components/layout/flow.rs, components/layout/fragment.rs, components/layout/table_rowgroup.rs, components/layout/table_wrapper.rs, components/layout/block.rs, components/layout/table_row.rs, components/style/matching.rs, components/layout/flow_list.rs, components/layout/table.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* This pull request modifies the contents of
`tests/wpt/css-tests/`, which are overwriten occasionally whenever the
directory is synced from upstream.

pcwalton commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 392ab81 with merge f930608...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3717)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13848) made this pull request unmergeable. Please resolve the merge conflicts.

Ms2ger commented:
r? @mbrubeck 

All I'd complain about is spec-related :)

emilio commented on components/layout/construct.rs:
```diff
@@ -1161,11 +1064,28 @@ impl<'a, ConcreteThreadSafeLayoutNode: ThreadSafeLayoutNode>
+        let table_style = node.style(self.style_context());
+        let wrapper_style = self.style_context()
+                                .stylist
+                                .precomputed_values_for_pseudo(&PseudoElement::ServoTableWrapper,
```
This pattern is becoming increasingly common, maybe add a helper for it in `SharedStyleContext`? Can be left as a followup, but please fill it if you decide to do so.

emilio reviewed (commented):
Just a first pass with a few comments. The style parts look good to me. The layout parts make sense to me, though I'd have to do a lot of spec-reading to ensure it's correct.

mbrubeck commented:
Some test results that weren't auto-linked above:

[mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3695)
```
  ▶ PASS [expected FAIL] /mediaqueries-3_dev/html/min-width-tables-001.htm
```

[linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/355)
<details>
```
Tests with unexpected results:
  ▶ PASS [expected FAIL] /css21_dev/html4/color-applies-to-014.htm

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-009.htm
  └   → /css21_dev/html4/table-anonymous-objects-009.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-020.htm
  └   → /css21_dev/html4/table-anonymous-objects-020.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-019.htm
  └   → /css21_dev/html4/table-anonymous-objects-019.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-010.htm
  └   → /css21_dev/html4/table-anonymous-objects-010.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-018.htm
  └   → /css21_dev/html4/table-anonymous-objects-018.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-012.htm
  └   → /css21_dev/html4/table-anonymous-objects-012.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-011.htm
  └   → /css21_dev/html4/table-anonymous-objects-011.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-017.htm
  └   → /css21_dev/html4/table-anonymous-objects-017.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-197.htm
  └   → /css21_dev/html4/table-anonymous-objects-197.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-198.htm
  └   → /css21_dev/html4/table-anonymous-objects-198.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-199.htm
  └   → /css21_dev/html4/table-anonymous-objects-199.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-200.htm
  └   → /css21_dev/html4/table-anonymous-objects-200.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-203.htm
  └   → /css21_dev/html4/table-anonymous-objects-203.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-202.htm
  └   → /css21_dev/html4/table-anonymous-objects-202.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-201.htm
  └   → /css21_dev/html4/table-anonymous-objects-201.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ FAIL [expected PASS] /css21_dev/html4/table-anonymous-objects-204.htm
  └   → /css21_dev/html4/table-anonymous-objects-204.htm 9a467359e61bf1cd03e915a871942253e111a547
/css21_dev/html4/reference/no_red_antialiasing_a_bc_d-ref.htm 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1
Testing 9a467359e61bf1cd03e915a871942253e111a547 == 398e25f2937dcbd574d5e6d4aeb2542b5e1752f1

  ▶ PASS [expected FAIL] /mediaqueries-3_dev/html/min-width-tables-001.htm
```
</details>

pcwalton commented:
I've investigated the failures.

* `/css21_dev/html4/color-applies-to-014.htm`: As far as I can tell, this is an incorrect test because it fails when browsers perform pixel snapping. Specifically, it tests to make sure that (effectively) `<div>Filler Text</div>` renders identitally to `<table><tr><td>Filler</td><td>&nbsp;Text</td></tr></table>` (with all borders/padding/margins/spacing on the tables removed). This can fail in browser engines like Servo that perform pixel snapping on table cells, since the "T" glyph can be subpixel aligned in the former case but not the latter.

* `/css21_dev/html4/table-anonymous-objects-???.htm`: These tests rely on summing a set of fractional pixel quantities to overlay the green text "abcd" directly on top of red text "abcd". They then check that the result is precisely equal to the green text "abcd". *The display lists come out correct from layout*, but WebRender introduces a small amount of rounding error somewhere on Linux (only) and there end up being tiny one-pixel differences around the edges of the letters.

I propose:

1. Ignoring `color-applies-to-014.htm` outright, as it's an incorrect test.

2. Marking `table-anonymous-objects-???.htm` as failing on Linux, as they trigger minor WebRender bugs on that platform.

Thoughts?

Ms2ger commented:
As long as you mean "mark as failing" rather than "ignore".

pcwalton commented:
What's the benefit of marking a test we believe to be invalid as failing instead of ignoring it? Its success depends on the details of the font on the platform.

pcwalton commented on components/layout/construct.rs:
```diff
@@ -1923,3 +1849,189 @@ impl ComputedValueUtils for ServoComputedValues {
+            }
+
+            (FlowClass::Flex, FlowClass::Block) |
+            (FlowClass::Flex, FlowClass::Flex) => {
```
If I don't allow `FlowClass::Flex` it breaks all the flexbox tests. This patch is too big already. I don't want to rewrite all the flexbox code.

stshine replied:
Oh, I mean IMO the wrapper block should only be needed for `FlowClass::Inline` since other children are guaranteed to be block-like.

pcwalton commented:
Rebased and addressed review comments. Next steps: address test failures and split up the patch.

pcwalton committed:
layout: Refactor `Flow::from_fragment`-style constructors to be
consistent.

pcwalton committed:
layout: Add a new `MARGINS_CANNOT_COLLAPSE` flow flag.

pcwalton committed:
layout: Treat flex flows as individual formatting contexts per spec.

pcwalton committed:
layout: Make `surrounding_intrinsic_inline_size` report border/padding
and margin separately.

This makes correct table layout easier.

pcwalton committed:
layout: Don't leak `FlowRef`s to the flexbox layout code.

This can cause memory safety problems. Comments have been added to
`flow_list.rs` to prevent this from happening again.

pcwalton committed:
style: Add some new Servo-specific pseudo-classes for anonymous flows.

This allows us to remove `modify_style_for_anonymous_table_object`.

pcwalton committed:
layout: Add a new cascading mode that inherits all properties, even
non-inheritable ones.

This works like the `modify_style_for_*` functions and will allow us to
easily migrate from them to real cascading.

pcwalton committed:
resources: Style anonymous blocks and anonymous table flows
appropriately in the UA stylesheet.

pcwalton committed:
layout: Don't use `BlockFlow::assign_block_size()` when assigning flex
flows' block sizes.

This ensures that we never collapse margins for flex flows.

pcwalton committed:
layout: Rewrite the anonymous table object generation code.

This patch introduces a "legalizer", which encapsulates all the
anonymous flow generation logic in one place. The legalizer knows how
to, for example, place adjacent blocks with `display: table-cell` in the
same table row.

Improves etsy.com.

Closes #13782.

pcwalton committed:
layout: Remove the old anonymous table object generation code.

pcwalton committed:
style: Remove `modify_style_for_anonymous_table_object`, now that the
cascade takes care of it.

pcwalton committed:
layout: Implement `TableWrapper::positioning()`.

The base `Flow` implementation returns the incorrect value.

pcwalton committed:
layout: Use the new `style_for_anonymous_box()` function for the
anonymous text child of `<input>`.

pcwalton committed:
layout: Assign border and padding to tables instead of table wrappers.

This is more straightforward when anonymous table object generation is
involved.

pcwalton committed:
layout: Make intrinsic inline sizes take the value of `box-sizing` into
account.

pcwalton committed:
tests: Update test expectations.

pcwalton committed:
layout: Minor style cleanup.

pcwalton commented:
OK, comments addressed and patch cleaned up. This should be ready for review.

r? @mbrubeck

stshine commented:
@bors-servo try

#13900 is merged.

bors-servo commented:
:hourglass: Trying commit b2e7831 with merge c3aba93...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/3759)

emilio commented on components/style/selector_matching.rs:
```diff
@@ -252,25 +253,62 @@ impl Stylist {
     }
 
+    /// Returns the style for an anonymous box of the given type.
+    pub fn style_for_anonymous_box(&self,
```
You should mark this function as `#[cfg(feature = "servo")]` so this compiles on Gecko.

pcwalton commented:
@emilio Done.

emilio commented on components/script/dom/storage.rs:
```diff
@@ -196,7 +196,7 @@ impl Runnable for StorageEventRunnable {
         for it_context in root_context.iter() {
             let it_window = it_context.active_window();
-            assert!(UrlHelper::SameOrigin(&ev_url, &it_window.get_url()));
+            //assert!(UrlHelper::SameOrigin(&ev_url, &it_window.get_url()));
```
Is this comment intentional?

pcwalton replied:
Argh, no. I'll remove it. I keep having to make this change locally because this assertion fires all the time when saving pages locally to disk.

emilio commented on components/layout/flow_list.rs:
```diff
@@ -52,27 +59,26 @@ impl FlowList {
     /// Provide a forward iterator with mutable references
+    ///
+    /// SECURITY-NOTE(pcwalton): This does not hand out `FlowRef`s by design. Do not add a method
+    /// to do so! See the comment above in `FlowList`.
```
++

emilio commented on components/layout/flow_list.rs:
```diff
@@ -93,6 +99,21 @@ impl FlowList {
 }
 
+impl Index<usize> for FlowList {
+    /// FIXME(pcwalton): O(n)!
```
:(.

This is only used for flexbox for now right? I wonder if we could rename it to `nth_slow` and `nth_slow_mut` so it's more obvious and so it's not misused without knowledge.

emilio reviewed (commented):
The layout code makes also sense to me, though I'd need to read a lot more about anonymous table wrappers so I can confidently r+ it.

If nobody reviews it this week I guess I'll do...

pcwalton committed:
layout: Add a safe way for layout to perform random access on child
flows.

This uses caching to ensure that we perform no more than O(n) pointer
loads.

emilio commented:
@pcwalton: @mbrubeck said he was going to take a look at this today, but if he doesn't this looks good to me.

jdm commented:
`./mach build-geckolib` is failing, as is `manifest_changed.sh`.

jdm commented on tests/wpt/mozilla/meta/MANIFEST.json:
```diff
@@ -1424,6 +1424,18 @@
             "url": "/_mozilla/css/dirty_viewport.html"
           }
         ],
+        "css/display_table_cell_intrinsic_width_a.html": [
```
The new reftest appears to be missing from this PR.

pcwalton commented:
Both issues should be fixed now.

mbrubeck commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit ba005a9 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck ba005a93b89c5b6b1a4ce315c5d30058f7bfae1e -->

bors-servo commented:
:hourglass: Testing commit ba005a9 with merge 6b40f97...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3722), [arm64](http://build.servo.org/builders/arm64/builds/3733), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3804), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/443), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/442), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3802), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3789), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/482), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/483), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1705)
